juju/model.py

```diff
@@ -0,0 +1,210 @@
+        return [Relation(relation_name, relation_id, self._local_app, self._local_unit, self._backend) for relation_id in relation_ids]
+
+
+class Relation:
+    def __init__(self, relation_name, relation_id, local_app, local_unit, backend=None):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend or ModelBackend()
+        self.data = RelationData(self.relation_name, relation_id, local_app, local_unit, self._backend)
+
+    @property
+    @lru_cache(maxsize=None)
+    def apps(self):
+        # Lazy-load the apps on this relation.
+        unit_names = self._backend.relation_list(self.relation_id)
+        apps = {}
+        for unit_name in unit_names:
+            app_name = unit_name.split('/')[0]
+            app = apps.get(app_name)
+            if not app:
+                app = apps[app_name] = Application(app_name, [])
+            app.units.append(Unit(unit_name))
+        return list(apps.values())
+
+
+class RelationData(Mapping):
+    def __init__(self, relation_name, relation_id, local_app, local_unit, backend=None):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._local_app = local_app
+        self._local_unit = local_unit
+        self._backend = backend or ModelBackend()
+
+    def __hash__(self):
+        return id(self)
+
+    def __eq__(self, other):
+        return self is other
+
+    @property
+    @lru_cache(maxsize=None)
+    def _members(self):
+        # Lazy-load the list of members for this relation.
+        members = set()
+        members.add(self._local_app)
+        members.add(self._local_unit)
+        unit_names = self._backend.relation_list(self.relation_id)
+        apps = {}
+        for unit_name in unit_names:
+            app_name = unit_name.split('/')[0]
+            app = apps.get(app_name)
+            if not app:
+                app = apps[app_name] = Application(app_name, [])
+                members.add(app)
+            unit = Unit(unit_name)
+            members.add(unit)
+            app.units.append(unit)
+        return members
+
+    def __iter__(self):
+        return iter(self._members)
+
+    def __len__(self):
+        return len(self._members)
+
+    @lru_cache(maxsize=None)
+    def __getitem__(self, member):
+        # Lazy-load the data for this relation member.
+        if member not in self._members:
+            raise KeyError(member)
+        return self._backend.relation_get(self.relation_id, member.name)
+
+
+class ModelError(Exception):
+    pass
+
+
+class TooManyRelatedApps(ModelError):
+    def __init__(self, relation_name, num_related, max_supported):
+        super().__init__(f'Too many remote applications on {relation_name} ({num_related} > {max_supported})')
+        self.relation_name = relation_name
+        self.num_related = num_related
+        self.max_supported = max_supported
+
+
+class ModelBackend:
+    def goal_state(self):
```

Comment:
Can we please drop this for now to make sure it's not used?  Happy to discuss ideas to solve the problems that it addresses when the time comes.

Reply:
I think we have to address this now if we want `model.app.units` to be populated at all, and I don't see another option currently but I'm certainly open to ideas.